xpdtools/calib2.py

```diff
@@ -0,0 +1,209 @@
+
+import numpy as np
+
+import tifffile as tf
+import matplotlib as mpl
+
+mpl.use('TkAgg')
+import matplotlib.pyplot as plt
+
+
+filename = 'Ni_pin_20181101-075909_973de2_0001.tiff'
+
+imarray=tf.imread(filename)
+print(imarray)
+
+class Slyce():
+	def __init__(self,direction,index,imarray):
+		self.direction=direction
+		self.index=index
+		self.pixels=[]
+		if direction=='v':
+			self.data=list(imarray[:,index])
+		if direction=='h':
+			self.data=list(imarray[index,:])
+
+def findringcenter(image,thres=.20,d=20):
+	
+	def findcenter(image):
+		s=image.shape
+		# number of rows divided by 2
+		r=s[0]/2
+		# number of columns divided by 2
+		c=s[1]/2
+		return r,c
+
+	r,c=findcenter(image)
+
+	slyce1=Slyce('v',c-d,image)
+	slyce2=Slyce('v',c+d,image)
+	slyce3=Slyce('v',c,image)
+	slyce4=Slyce('v',c-d/2,image)
+	slyce5=Slyce('v',c+d/2,image)
+
+	slyce6=Slyce('h',r-d,image)
+	slyce7=Slyce('h',r+d,image)
+	slyce8=Slyce('h',r,image)
+	slyce9=Slyce('h',r+d/2,image)
+	slyce10=Slyce('h',r-d/2,image)
+	slyces=[slyce1, slyce2, slyce3, slyce4, slyce5, slyce6, slyce7, slyce8, slyce9, slyce10]
+
+	def findthreshold(image,thres):
+		return thres*np.max(image)
+
+	thresh=findthreshold(image,thres)
+	
+	def clickpoints(slyce,thresh):
+		p1=np.argmax(slyce.data)
+		p2=np.argmax(slyce.data[:p1]+slyce.data[p1+1:])
+
+		if slyce.data[p2]>=thresh:
+			slyce.pixels.append(p1)
+			slyce.pixels.append(p2)
+		
+		elif slyce.data[p1]>=thresh:
+
+			slyce.pixels.append(p1)
+	
+	points2click=[]
+	for slyce in slyces:
+		clickpoints(slyce,thresh)
+		if slyce.direction=='v':
+			for pixel in slyce.pixels:
+				points2click.append([pixel,slyce.index])
+		if slyce.direction=='h':
+			for pixel in slyce.pixels:
+				points2click.append([slyce.index,pixel])
+
+	highlight=np.mean(image)*.9
+	image_new=image
+	
+	def finddistance(a,b):
+		dif1=a[1]-b[1]
+		dif2=a[0]-b[0]
+		d_sq=np.abs(dif1**2+dif2**2)
+		return np.sqrt(d_sq)
+
+	
+	#f, ax = plt.subplots(1)
+	#ax.imshow(image)
+	ring_mask=np.ones_like(image,dtype=bool)
+	
+
+	for point in points2click:
+		#ax.scatter(point[0],point[1])
+		ring_mask[point[0],point[1]]=False
+		image_new[point[0],point[1]]=0
+	
+	coords=[]
+	spread=[]
+	for row in range (int(r-3*d),int(r+3*d)):
```

Comment:
I don't know if I understand how this works.

Reply:
The for loop goes through points reasonably close to the center of the image (not necessary the rings which could pose a problem in some cases). It them finds the distance between that point and all the points that had been identified as being on the center ring. The point that is closest to being equidistant from the points on the inner ring is identified as the center point.